src/cpp/rtps/reader/timedevent/HeartbeatResponseDelay.cpp

```diff
@@ -70,8 +70,10 @@ void HeartbeatResponseDelay::event(
     {
         logInfo(RTPS_READER,"");
 
-        // Protect reader
-        std::lock_guard<std::recursive_timed_mutex> guard(mp_WP->mp_SFR->getMutex());
+       // Protect reader
+       // std::lock_guard<std::recursive_timed_mutex> guard(mp_WP->mp_SFR->getMutex());
```

Comment:
This mutex should be locked here.

Reply:
When we promoted PDP endpoints to reliable we introduced an unnoticed synchronization problem. 
When a DATA(p[UD]) is received the server unmatches its PDP endpoints (as PDPSimple does) but those are now reliable. This implies the destruction of the remote participant WriterProxies. 
There is a particular WriterProxy event (HeartbeatResponseDelay) that locks on PDP reader mutex in order to send some messages. 
The ~WriterProxy waits for the event thread if its busy before destroying it.

If we are unlucky and the HeartbeatResponseDelay is running during our DATA(p[UD]) callback a deadlock will ensue:
+ The event thread will wait till the DATA(p[UD]) callback unlocks PDP Reader mutex.
+ The DATA(p[UD]) callback waits for the event thread to finish.

In order to solve this problem I modify PDPServer to unlock the PDPReader during the unmatch operation allowing the event thread to gracefully finish.